Email Notice when Export Download is Ready

xbryanc opened the pull request:
Added a button that emails the user when his/her export is ready to download. Once the user has clicked the button and the request has been sent to the server, the user may close the window since the rest is taken care of server-side.

**This was done using a VM, and absolute_reverse caused a hang when clicking the button. This still must be tested on a computer with sufficient memory @dannyroberts **

Edit: Verified to work on a regular computer

xbryanc committed:
add 'email me when this is done' button to exports page

Button currently only creates an EmailRequest object

xbryanc committed:
rename EmailRequests to EmailRequest

Also deleted extraneous debugging

xbryanc committed:
add email functionality to EmailRequest

Implemented addition, searching, and removal of the EmailRequest database, and connected these to the email functionality of the export email button. Still need to test if absolute_reverse works or hangs on a computer with sufficient memory.

xbryanc committed:
fix lack of dangling comma in ajax

Thanks sticklerbot

orangejenny commented on corehq/apps/export/static/export/js/download_export.ng.js:
```diff
@@ -286,6 +287,21 @@
+                    $.ajax({
+                        method: 'POST',
+                        dataType: 'json',
+                        url: '../../../../../add_export_email_request/',
```
There's infrastructure for providing URLs to JavaScript without needing to hard code them. Check out https://github.com/dimagi/js-guide/blob/master/integration-patterns.md#django-urls

dannyroberts replied:
+1 to this. @xbryanc I know I showed you how to do this so you could get over the hump, but definitely worth fixing to have this follow that practice. (Keeps links from breaking if we change URLs, etc., but also is just standard practice that helps others follow the code.)

orangejenny commented on corehq/apps/export/migrations/0004_emailrequest_email_address.py:
```diff
@@ -0,0 +1,21 @@
+        migrations.AddField(
+            model_name='emailrequest',
+            name='email_address',
+            field=models.CharField(default='foo@gmail.com', max_length=255),
```
Odd default - why not blank string?

orangejenny commented on corehq/apps/export/templates/export/partial/export_download_progress.html:
```diff
@@ -52,8 +52,19 @@
                         </button>
+                        {% if request|toggle_enabled:"EMAIL_EXPORT_WHEN_DONE_BUTTON" %}
+                        <button type="button"
+                                class="btn btn-primary btn-full-width"
```
`btn-primary` is only for a single primary action on a page. I'd switch this to `btn-default`.

orangejenny commented on corehq/ex-submodules/soil/util.py:
```diff
@@ -88,6 +92,20 @@ def get_download_context(download_id, message=None, require_result=False):
+    except TaskFailedError:
+        allow_dropbox_sync = False
+    download_url = "{}?get_file".format(absolute_reverse('retrieve_download', args=(download_id,)))
+    email_body = 'Your CommCare export is ready! Click on the link below to download your ' + \
```
translate?

orangejenny replied:
same with the subject line below

dannyroberts replied:
@xbryanc "translating" here means marking for translation. Every bit of text we show the user should be marked for translation, so that when a user who speaks French sees it, they see it in french. (In this particular case, there may be a bit more we need to do to make this user's language preference available at this point in the code, but we can handle that down the road—at least with it marked for translation our translators can translate it in the meantime.)

Marking for translation is explained here: https://docs.djangoproject.com/en/1.11/topics/i18n/translation/. In addition, the use of the `+` operator for string concatenation is strongly discouraged in python; use `'My favorite food is {}!'.format('bananas')`-style string formatting instead.

orangejenny commented on corehq/apps/export/views.py:
```diff
@@ -2481,6 +2482,23 @@ def can_download_daily_saved_export(export, domain, couch_user):
     return False
 
 
+@login_and_domain_required
```
Might as well also use `@require_POST`

orangejenny commented on corehq/apps/export/views.py:
```diff
@@ -2481,6 +2482,23 @@ def can_download_daily_saved_export(export, domain, couch_user):
+    except TaskFailedError:
+        return HttpResponseServerError()
+    if download_id is None or email_address is None:
+        return HttpResponseBadRequest()
```
Error message would be nice.

orangejenny commented on corehq/apps/export/templates/export/partial/export_download_progress.html:
```diff
@@ -52,8 +52,19 @@
+                        {% endif %}
                         {% if show_sync_to_dropbox %}
-                        <button button="submit"
+                        <button type="submit"
```
Curious, why was this change necessary?

dannyroberts replied:
This confused us, but is just a random change. Is `<button button="submit">` normal/valid html? I assumed it was a typo.

orangejenny replied:
The original code does look like a typo, but I had thought that the default button type was `button` and switching to `submit` could change behavior. But I just looked it up and learned the [default is submit](https://stackoverflow.com/questions/31644723/what-is-the-default-button-type), so yeah, should be a fine change.

dannyroberts replied:
Ha! Thanks for digging. That's kinda funny, but makes sense. This worked for the person who wrote it but not for the reason that they thought it did.

orangejenny reviewed (commented):
Didn't read closely, just skimmed for best practices.

Seems like a good feature in general, great for those with low connectivity. @dannyroberts is there a plan to get it un-feature-flagged?

orangejenny commented:
> This was done using a VM, and absolute_reverse caused a hang when clicking the button.

If responsiveness is a concern - and we probably have users who have machines with as little memory as your VM - it'd be good to use [disableButton](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/hq.helpers.js#L138-L141) which'll add a spinner to the button so the user doesn't think the page is just not responding.

dannyroberts commented on corehq/apps/export/models/new.py:
```diff
@@ -2675,6 +2675,12 @@ def delete(self, using=None, keep_parents=False):
+class EmailRequest(models.Model):
+    domain = models.CharField(max_length=255)
+    download_id = models.CharField(max_length=255)
+    email_address = models.CharField(max_length=255)
```
While we may only need an email address here, I think it would be better to put the user's `user_id` and then use that to look up the user and finally get the email address once you're in the task. There are a couple reasons for this. First, the user is the richer context from which you can derive the `email_address`—if later on we need to know other things about the user (like their language preference for instance) it will be nice to have available.

If you don't change this now, that's not the end of the world, but if it's not too hard, I think it'd be a minor improvement.

dannyroberts commented on corehq/apps/export/static/export/js/download_export.ng.js:
```diff
@@ -286,6 +287,21 @@
         };
+
+        $scope.sendEmailUponCompletion = function () {
+            setTimeout(function () {
```
I would add a comment here about why you're doing this in an asynchronous loop. It may not be clear to the reader that it's to wait until `download_id` is available in the scope.

dannyroberts reviewed (commented):
Looking good. Left a few comments

orangejenny commented:
@xbryanc Danny gently pointed out that my comment about our users' machines makes no sense, feel free to ignore.

xbryanc committed:
change EmailRequest to EmailExportWhenDoneRequest, fix client to server requests, make code and display changes

Various fixes:
-Changed name of database and reversed prior migrations
-No longer stores email address of user, now more broadly stores his/her id
-Changed button HTML
-Normalized the way the client requests the server
-Added translation support
-Various formatting changes and comment additions

xbryanc committed:
fix maximum line length

Moved part of an import statement to another line, thanks SticklerBot

dannyroberts commented on corehq/ex-submodules/soil/util.py:
```diff
@@ -94,16 +95,18 @@ def get_download_context(download_id, message=None, require_result=False):
+    if allow_dropbox_sync:
+        dropbox_message = ('\n\nYou can also upload your data to Dropbox with the link below:\n'
+                           '{}').format(dropbox_url)
+    email_body = ('Your CommCare export is ready! Click on the link below to download your requested data:\n'
```
I think you didn't actually translate these? You need to use `_` the way you did below.

xbryanc committed:
fix translation marking for email message

Moved translation marking to the string literals as it should be
